storage/rocksdb/rdb_compact_filter.h

```diff
@@ -50,15 +110,13 @@ class Rdb_compact_filter : public rocksdb::CompactionFilter {
   void calculate_expiration_timestamp() const {
     assert(!m_expiration_timestamp.has_value());
 
-    uint64_t oldest_snapshot_timestamp = 0;
-    rocksdb::DB *const rdb = rdb_get_rocksdb_db();
-    if (!rdb->GetIntProperty(rocksdb::DB::Properties::kOldestSnapshotTime,
-                             &oldest_snapshot_timestamp) ||
-        oldest_snapshot_timestamp == 0) {
-      oldest_snapshot_timestamp = static_cast<uint64_t>(std::time(nullptr));
+    uint64_t oldest_transaction_timestamp =
+        Rdb_transaction_timestamps::instance().oldest();
+    if (oldest_transaction_timestamp == 0) {
+      oldest_transaction_timestamp = static_cast<uint64_t>(std::time(nullptr));
     }
 
-    m_expiration_timestamp = oldest_snapshot_timestamp;
+    m_expiration_timestamp = oldest_transaction_timestamp;
```

Comment:
Hey Laurynas, we recently notice a new issue that will bring inconsistency to primary and secondary in long running transactions. The oldest_transaction_timestamp here is retrieved from snapshot->GetUnixTime(). This timestamp will be different between primary and secondary. Secondary will have a slightly larger timestamp, hence compact more data away than primary. Can we change the oldest_transaction_timestamp from snapshot->GetUnixTime() to mysql_bin_log.get_max_write_hlc_nsec()? This can ensure the consistency between primary and secondary because mysql_bin_log.get_max_write_hlc_nsec() returns the same value if primary and secondary has the same GTID set when taking snapshot.
cc @abhinav04sharma

Reply:
#1471 has been closed. Can we have a rebase to include that change? Thanks!

Reply:
@sunshine-Chun, OK, I will make it use `mysql_bin_log.get_max_write_hlc_nsec()`. I am rebasing (and rewriting) this PR right now.

Reply:
@sunshine-Chun , I am not too familiar with the binlog HLC patch. What value should we use if the HLC value is zero? Should we use `snapshot->GetUnixTime()` then?

Reply:
@sunshine-Chun, the relationship between binlog and RocksDB timestamps is not clear to me. We initialize the earliest timestamp from binlog, but then the current snapshot timestamps come from the RocksDB snapshots directly. I'd expect the invariant "the current timestamp can be earlier than the earliest transaction timestamp" to hold, but, i.e. `rocksdb_rpl.rocksdb_binlog_ttl` test shows (`result` is the current snapshot timestamp)
```
(lldb) p/x m_earliest_snapshot_ts
(int64_t) 0x00047fc94dc5cd7c
(lldb) p/x result
(const int64_t) 0x0000000066c7143b
```

Reply:
Here is what these two timestamp represents:
1. Each write transaction has a HLC timestamp assigned to it. The `mysql_bin_log.get_max_write_hlc_nsec()` tells us the lastest HLC that is written to this instance. 
2. The RocksDB timestamp represents the unix time when creating the snapshot. 

In most cases, these two timestamp should be very close in value. In the long running transaction case, when creating snapshot,  `mysql_bin_log.get_max_write_hlc_nsec()` should always be smaller than the RocksDB timestamp, cause no new writes will update the max_write_hlc_nsec at that time.

The reason why we use  `mysql_bin_log.get_max_write_hlc_nsec()` here is that it can help the primary and secondary be consistent at the this value. When we create snapshot in long running transactions in primary and secondary, we can make sure they are on the same GTID set. This also means they will have the same `mysql_bin_log.get_max_write_hlc_nsec()` value. Noted primary and secondary maintains their own `mysql_bin_log.get_max_write_hlc_nsec()` value. But since they have the same committed transactions, they share the same `mysql_bin_log.get_max_write_hlc_nsec()` value. Rocksdb timestamp can not do this cause it will get the unix now in primary and secondary separately. Secondary takes snapshot later than primary.

Reply:
`mysql_bin_log.get_max_write_hlc_nsec()` is unlikely to be zero. But yeah, `snapshot->GetUnixTime()` seems the only thing we can do to guard the `new_earliest_ts` from zero.

Reply:
MTR test seems to be failing. Have you rebased it to the lastest branch?

Reply:
@sunshine-Chun, thank you, this makes sense. It seems that the "L" in "HLC" is not stored explicitly, thus the binlog HLC is directly comparable to RocksDB timestamps, different resolutions aside. In which case I still cannot understand what is going on with `rocksdb_rpl.rocksdb_binlog_ttl`. The testcase does not inject any artificial timestamps, and yet:
```
$ p result # The current RDB snapshot timestamp
(const int64_t) 1724752351 # OK, resolves to $NOWish timestamp at the second resolution
(lldb) p m_earliest_snapshot_ts # Binlog HLC
(int64_t) 1695877370630958 # Does not resolve to anything reasonable at any resolution
```

Reply:
I run the test in my local server and got the following number. The `m_earliest_snapshot_ts` looks correct to me. Note, this is the value that hit the assertion `assert(m_earliest_snapshot_ts <= result || result == 0);`
```
(gdb) p m_earliest_snapshot_ts
$1 = 1724827776952667189
(gdb) p result
$2 = 1724827777
```
Let me post the logic to get HLC here.
```
uint64_t HybridLogicalClock::get_next() {
  uint64_t current_hlc, next_hlc = 0;
  bool done = false;

  while (!done) {
    // Get the current wall clock in nanosecond precision
    uint64_t current_wall_clock =
        std::chrono::duration_cast<std::chrono::nanoseconds>(
            std::chrono::high_resolution_clock::now().time_since_epoch())
            .count();

    // get the 'current' internal HLC
    current_hlc = current_.load();

    // Next HLC timestamp is max of current-hlc and current-wall-clock
    next_hlc = std::max(current_hlc + 1, current_wall_clock);

    // Conditionally update the internal hlc
    done = current_.compare_exchange_strong(current_hlc, next_hlc);
  }

  return next_hlc;
}
```

Can try to print the `current_wall_clock` to see if the value is abnormal? This determines the baseline value of the HLC. My guess is that the abnormal value starts from here.

Reply:
`current_wall_clock` value is indeed correct, thus the only issue was comparing nanoseconds to seconds directly. Adding the required conversion fixed the tests.